### PR TITLE
Ruby: some improvements

### PR DESCRIPTION
**Pull request by aibaars**
I investigated a case where we didn't flag up a [result](https://github.com/rubygems/bundler/blob/674966e21cc7257dd52797ef56ee9628c3ce1bc6/lib/bundler.rb#L187) for the rb/overly-permissive-file query .
This pull request implements the improvements needed to flag up that result, and makes the following changes:

* track Pathname instances using global flow instead of local flow
* model Object.dup and Kernel.tap
* add flow steps for logical operators
* add support for extend in the call graph

The dataflow improvements caused a fairly large number of false positives on our test suite for the missing regex anchor and incomplete hostname regex queries. This was caused by misclassifying some string literals as regular expression due to confusing `Regexp#match` and `String#match`. Commit f93bc900b90c264c935b25d5eb669aa9e353b3c4 improves this and remove the false positive results.

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll [hunk 1]
```diff
@@ -181,6 +181,12 @@ module LocalFlow {
         ) and
         nodeFrom.asExpr() = for.getValue()
       )
+    or
+    nodeTo.asExpr() =
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:185]: This is already included in taint tracking, and I don't think it is safe to include in data flow, as the value may not be preserved.

Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:185]: Why wouldn't it be value preserving?

Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:185]: As far as I know:
* `a || b`  means `if (tmp = a) then tmp else b`
* `a && b`  means `if (tmp = a) then b else tmp`

Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:185]: You are right!

Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:185]: I guess I should remove the equivalent code from the taint steps.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll [hunk 2]
```diff
@@ -674,6 +721,11 @@ private predicate singletonMethodOnModule(MethodBase method, string name, Module
   )
   or
   flowsToSingletonMethodObject(trackModuleAccess(m), method, name)
+  or
+  exists(Module other |
+    extendCallModule(m, other) and
+    method = lookupMethod(other, name)
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:727]: Should this be limited to public methods?

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll [hunk 3]
```diff
@@ -708,6 +765,12 @@ predicate singletonMethodOnInstance(MethodBase method, string name, Expr object)
   not selfInModule(object.(SelfVariableReadAccess).getVariable(), _) and
   // cannot use `trackModuleAccess` because of negative recursion
   not exists(resolveConstantReadAccess(object))
+  or
+  exists(DataFlow::ExprNode receiver, Module other |
+    extendCall(receiver, other) and
+    object = receiver.getExprNode().getExpr() and
+    method = lookupMethod(other, name)
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:772]: Should this be limited to public methods?

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll (hunk 2 above)
Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:727]: Good point. `private` can be called, but only if the receiver is `self` and `protected` methods can be called if the receiver is a subclass of the enclosing class or something.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll [hunk 4]
```diff
@@ -181,6 +181,12 @@ module LocalFlow {
...
+      any(CfgNodes::ExprNodes::BinaryOperationCfgNode op |
+        op.getExpr() instanceof BinaryLogicalOperation and
+        nodeFrom.asExpr() = op.getAPredecessor()
```
Comment (aibaars) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:188]:
```suggestion
        nodeFrom.asExpr() = op.getAnOperand()
```

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll [hunk 5]
```diff
@@ -294,6 +294,30 @@ private DataFlowCallable viableLibraryCallable(DataFlowCall call) {
   )
 }
 
+/** Holds if there is a call like `receiver.extend(M)`. */
+pragma[nomagic]
+private predicate extendCall(DataFlow::ExprNode receiver, Module m) {
+  exists(DataFlow::CallNode extendCall |
+    extendCall.getMethodName() = "extend" and
+    exists(DataFlow::LocalSourceNode sourceNode | sourceNode.flowsTo(extendCall.getArgument(0)) |
```
Comment (aibaars) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:302]:
```suggestion
    exists(DataFlow::LocalSourceNode sourceNode | sourceNode.flowsTo(extendCall.getArgument(_)) |
```

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll (hunk 3 above)
Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll:772]: Same as above.

**Review by hvitved — APPROVED** [review]

**Pull request merged**